### PR TITLE
feat: Make downloading of generated pdf proposals easier

### PR DESCRIPTION
**Pull request by mutambaraf**
## Description
Make it easier for a developer to download PDFs for proposal(s) using an API token.

## Motivation and Context
We want to make PDFs linkable from outside the proposals system.

## How Has This Been Tested
- manual tests
- e2e tests

## Close

https://github.com/UserOfficeProject/user-office-project-issue-tracker/issues/751
https://github.com/UserOfficeProject/user-office-project-issue-tracker/issues/716

## Changes

- Add token access to factory middleware, so as to allow API tokens to be used for downloading proposals pdf.
- Implement Token access on `/pdf/proposal/proposal_pks` route and add query string filter `/pdf/proposal/id/proposal_ids?filter=id` for downloading pdf by ids
- Implement token access logging in factory middleware.
- Implement users token authentication and validation to download a given proposal(s).
- Add API Token failure logging with id and reason for the failure.
- Add [Other Services] tab where users can assign API Token permission for downloading proposals pdfs.

Files changed:
- apps/user-office-backend/src/buildContext.ts
- apps/user-office-backend/src/context/index.ts
- apps/user-office-backend/src/factory/pdf/genericTemplates.ts
- apps/user-office-backend/src/factory/pdf/proposal.ts
- apps/user-office-backend/src/factory/pdf/sample.ts
- apps/user-office-backend/src/middlewares/factory.ts
- apps/user-office-backend/src/middlewares/factory/pdf.ts
- apps/user-office-backend/src/queries/AdminQueries.ts
- apps/user-office-backend/src/resolvers/queries/GetAllQueryAndMutationMethodsQuery.ts
- apps/user-office-frontend/src/components/settings/apiAccessTokens/CreateUpdateApiAccessToken.tsx
- apps/user-office-frontend/src/context/DownloadContextProvider.tsx
- apps/user-office-frontend/src/graphql/admin/getAllQueriesAndMutations.graphql
- apps/user-office-frontend/src/hooks/admin/useQueriesAndMutationsData.ts

## Tests included/Docs Updated?

- [x] I have added tests to cover my changes.
- [ ] All relevant doc has been updated

### CODE REVIEW

File: apps/user-office-backend/src/factory/pdf/proposalTokenAccess.ts [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    );
+  }
+
+  private setAllTopicsInformation(out: ProposalPDFData): void {
+    // Information from each topic in proposal
+    const questionarySteps = this.proposalPDFData.questionarySteps;
+    if (questionarySteps)
+      for (const step of questionarySteps) {
+        if (!step) {
+          logger.logError('step not found', { ...questionarySteps }); // TODO: fix type of the second param in the lib (don't use Record<string, unknown>)
```
Comment (ACLay) [inline apps/user-office-backend/src/factory/pdf/proposalTokenAccess.ts:238]: Is this TODO something that still needs attention? If so, could it be removed and have a separate issue opened?

Comment (mutambaraf) [reply apps/user-office-backend/src/factory/pdf/proposalTokenAccess.ts:238]: Created this issue https://github.com/UserOfficeProject/user-office-project-issue-tracker/issues/761 but removing the comment without fixing the underlining problem might not be the best decision.

File: apps/user-office-backend/src/factory/pdf/proposalTokenAccess.ts [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+  private async getQuestionarySteps(): Promise<QuestionaryStep[]> {
+    let questionarySteps = null;
+    const questionaryId = this.proposalPDFData.proposal?.questionaryId;
+    if (questionaryId) {
+      questionarySteps = await this.questionaryDataSource.getQuestionarySteps(
+        questionaryId
+      );
+    }
+    if (isRejection(questionarySteps) || questionarySteps == null) {
+      throw new Error('Could not fetch questionary');
```
Comment (jekabs-karklins) [inline apps/user-office-backend/src/factory/pdf/proposalTokenAccess.ts:157]: Should we add some logging here? Looks like there is a `rejection.reason` has a piece of valuable information that should be stored or returned

**Comment by mutambaraf** [comment]
Just tested again it is adding an additional blank page

**Comment by Scott-James-Hurley** [comment]
When I try to create an API key I get redirected to the main menu for ISIS on Authenticate. Is this behaviour intentional?

**Comment by mutambaraf** [comment]
> When I try to create an API key I get redirected to the main menu for ISIS on Authenticate. Is this behaviour intentional?
No, will check this also but creation of API keys is covered by some tests.

File: apps/user-office-backend/src/factory/pdf/proposal.ts [hunk 3]
```diff
@@ -254,3 +357,185 @@ export const collectProposalPDFData = async (
 
   return out;
 };
+
+export const collectProposalPDFDataTokenAccess = async (
```
Comment (ACLay) [inline apps/user-office-backend/src/factory/pdf/proposal.ts:361]: Forgive my lack of familarity with the new proposals system, but why do we need a separate `collectProposalPDFDataTokenAccess`? It looks like it has similar arguments, structure and response to the above `collectProposalPDFData`, and gets called from the same location, so I'm curious how they differ and why we need so much almost duplicate code

Comment (mutambaraf) [reply apps/user-office-backend/src/factory/pdf/proposal.ts:361]: They are different collectProposalPDFDataTokenAccess don't go through graphql  queries but get data directly from data sources.

**Review by Scott-James-Hurley — APPROVED** [review]

**Review by jekabs-karklins — APPROVED** [review]

**Pull request merged**